### DISCUSSION TITLE
OPC UA Server Roadmap

### DISCUSSION THREAD

**Discussion by birche** [gopcua/opcua#429] [Q&A]
Hi,
I've got the impression that this project is focusing on the client part. What's the state for the server part? Can I build something useful with what currently exists?

**Comment by magiconair** [comment]
Hi @birche, we would love to have a server and we started a couple of attempts. The last one was last summer where I've resurrected the original server branch from @dwhutchison. There is no functional server since we don't have an AddressSpace implementation. I think the basics might be a start for further exploration but you can't build a server yet. Having said that it would be awesome if someone would be interested in picking that up.

**Reply by birche** [reply]
Thanks @magiconair, I haven't considered how big this endeavor would be, but I'll consider look into it. Would be great to have a Golang alternative to the open62541.

**Reply by dwhutchison** [reply]
The entire server implementation as a whole is a big job, however, it's broken up into multiple smaller pieces.  As @magiconair  noted, the immediate hurdle right now is an AddressSpace implementation.  The AddressSpace forms the base on which the rest of the services are built upon (basically everything useful interacts with the AddressSpace). I've had a few attempts at creating it and it never turned out very well so I ended up scrapping them.  To be honest, I'm probably overthinking it and it just needs a simple, uncomplicated implementation to start with.  Once we can store and retrieve nodes, the various services can be implemented as we need them with updates and fine-tuning of the AddressSpace as we understand the problems and needs more clearly.

I still have an interest in working on the server implementation but my time has been extremely limited with additional work and family obligations and haven't felt that I'm able to dedicate the time needed to fully support the server.  If there are others that can help with the development, or at least come up with a decent design for the AddressSpace, I'd be able to put some time towards it.

**Comment by magiconair** [comment]
Hi @dwhutchinson, that is good to know. I’ve had the same experience with the AddressSpace but I think I’d like to give it another shot for a basic implementation. Would be nice to have some basic functionality since I think it’ll be easier to then get help from others.

**Comment by magiconair** [comment]
@birche I've picked this up again and there is some small momentum. The `server` branch starts up, the ProSys client connects and I've created a user address space working with the foundation that @dwhutchison layed 3 years ago. I'm tracking this in https://github.com/gopcua/opcua/discussions/659 if you want to help out
